File: tests/Aspire.Hosting.CodeGeneration.Elixir.ExTests/test/aspire_transport_test.exs
defmodule Aspire.TransportTest do
  use ExUnit.Case, async: false

  alias Aspire.CancellationToken
  alias Aspire.FakeHost
  alias Aspire.Handle
  alias Aspire.Transport

  @socket_env "REMOTE_APP_HOST_SOCKET_PATH"
  @token_env "ASPIRE_REMOTE_APPHOST_TOKEN"

  # ── Framing ───────────────────────────────────────────────────────────────

  test "frames a request with Content-Length byte count" do
    host = start_host()
    transport = start_transport(host)

    task = Task.async(fn -> Transport.ping(transport) end)
    assert %{"method" => "ping"} = FakeHost.await_request()

    raw = FakeHost.received_bytes(host)
    assert {length, body} = parse_frame(raw)
    assert length == byte_size(body)
    assert %{"jsonrpc" => "2.0", "method" => "ping", "params" => []} = JSON.decode!(body)
    assert is_integer(JSON.decode!(body)["id"])

    FakeHost.send_message(host, %{"jsonrpc" => "2.0", "id" => 1, "result" => "pong"})
    assert {:ok, "pong"} = Task.await(task)
  end

  test "parses two frames in one packet" do
    host = start_host()
    transport = start_transport(host)

    first = Task.async(fn -> Transport.invoke_capability(transport, "one", %{}) end)
    assert %{"id" => first_id} = FakeHost.await_request()

    second = Task.async(fn -> Transport.invoke_capability(transport, "two", %{}) end)
    assert %{"id" => second_id} = FakeHost.await_request()

    packet =
      FakeHost.frame(%{"jsonrpc" => "2.0", "id" => first_id, "result" => "first"}) <>
        FakeHost.frame(%{"jsonrpc" => "2.0", "id" => second_id, "result" => "second"})

    FakeHost.send_raw(host, packet)

    assert {:ok, "first"} = Task.await(first)
    assert {:ok, "second"} = Task.await(second)
  end

  test "parses a frame split across packets" do
    host = start_host()
    transport = start_transport(host)

    task = Task.async(fn -> Transport.invoke_capability(transport, "one", %{}) end)
    assert %{"id" => id} = FakeHost.await_request()

    frame = FakeHost.frame(%{"jsonrpc" => "2.0", "id" => id, "result" => "split"})
    <<head::binary-size(12), middle::binary-size(20), tail::binary>> = frame

    FakeHost.send_raw(host, head)
    refute Task.yield(task, 50)

    FakeHost.send_raw(host, middle)
    refute Task.yield(task, 50)

    FakeHost.send_raw(host, tail)
    assert {:ok, "split"} = Task.await(task)
  end

  test "multibyte UTF-8 body length is bytes not graphemes" do
    host = start_host()
    transport = start_transport(host)

    text = "héllo → 世界"
    task = Task.async(fn -> Transport.invoke_capability(transport, "echo", %{"text" => text}) end)
    assert %{"id" => id} = FakeHost.await_request()

    raw = FakeHost.received_bytes(host)
    assert {length, body} = parse_frame(raw)
    assert length == byte_size(body)
    assert length > String.length(body)
    assert %{"params" => ["echo", %{"text" => ^text}]} = JSON.decode!(body)

    FakeHost.send_message(host, %{"jsonrpc" => "2.0", "id" => id, "result" => text})
    assert {:ok, ^text} = Task.await(task)
  end

  # ── Requests ──────────────────────────────────────────────────────────────

  test "ping returns pong" do
    host = start_host(handler: method_handler(%{"ping" => "pong"}))
    transport = start_transport(host)

    assert {:ok, "pong"} = Transport.ping(transport)
    assert %{"method" => "ping", "params" => []} = FakeHost.await_request()
  end

  test "invoke_capability returns a handle" do
    result = %{
      "$handle" => "2",
      "$type" => "Aspire.Hosting.Redis/Aspire.Hosting.ApplicationModel.RedisResource"
    }

    host = start_host(handler: method_handler(%{"invokeCapability" => result}))
    transport = start_transport(host)

    builder = %Handle{
      id: "1",
      type: "Aspire.Hosting/Aspire.Hosting.IDistributedApplicationBuilder"
    }

    assert {:ok, %Handle{id: "2", type: type}} =
             Transport.invoke_capability(transport, "Aspire.Hosting.Redis/addRedis", %{
               "builder" => builder,
               "name" => "cache"
             })

    assert type == "Aspire.Hosting.Redis/Aspire.Hosting.ApplicationModel.RedisResource"

    assert %{"method" => "invokeCapability", "params" => params} = FakeHost.await_request()

    assert params == [
             "Aspire.Hosting.Redis/addRedis",
             %{
               "builder" => %{
                 "$handle" => "1",
                 "$type" => "Aspire.Hosting/Aspire.Hosting.IDistributedApplicationBuilder"
               },
               "name" => "cache"
             }
           ]
  end

  test "invoke_capability error maps to Aspire.Error with code and data" do
    error = %{
      "$error" => %{
        "code" => "CAPABILITY_NOT_FOUND",
        "message" => "Unknown capability: Contoso.Aspire/bar",
        "capability" => "Contoso.Aspire/bar",
        "details" => %{"parameter" => "name"}
      }
    }

    host = start_host(handler: method_handler(%{"invokeCapability" => error}))
    transport = start_transport(host)

    assert {:error, %Aspire.Error{} = returned} =
             Transport.invoke_capability(transport, "Contoso.Aspire/bar", %{})

    assert returned.code == "CAPABILITY_NOT_FOUND"
    assert returned.message == "Unknown capability: Contoso.Aspire/bar"
    assert returned.capability == "Contoso.Aspire/bar"
    assert returned.data == error["$error"]
    assert Exception.message(returned) == "Unknown capability: Contoso.Aspire/bar"
  end

  test "concurrent requests correlate by id" do
    host = start_host()
    transport = start_transport(host)

    tasks =
      for index <- 1..5 do
        Task.async(fn -> Transport.invoke_capability(transport, "cap#{index}", %{}) end)
      end

    requests = for _ <- 1..5, do: FakeHost.await_request()
    assert Enum.map(requests, & &1["id"]) == Enum.sort(Enum.map(requests, & &1["id"]))

    # Answer in reverse order to prove correlation is by id, not by arrival.
    for request <- Enum.reverse(requests) do
      ["cap" <> index, _args] = request["params"]

      FakeHost.send_message(host, %{
        "jsonrpc" => "2.0",
        "id" => request["id"],
        "result" => "result#{index}"
      })
    end

    assert Enum.map(tasks, &Task.await/1) ==
             for(index <- 1..5, do: {:ok, "result#{index}"})
  end

  test "cancel_token sends cancelToken" do
    host = start_host(handler: method_handler(%{"cancelToken" => true}))
    transport = start_transport(host)

    assert {:ok, true} = Transport.cancel_token(transport, "ct_abc")

    assert %{"method" => "cancelToken", "params" => ["ct_abc"]} = FakeHost.await_request()
  end

  test "Aspire.CancellationToken.cancel sends cancelToken" do
    host = start_host(handler: method_handler(%{"cancelToken" => true}))
    transport = start_transport(host)

    token = CancellationToken.new()
    assert {:ok, true} = CancellationToken.cancel(token, transport)

    assert %{"method" => "cancelToken", "params" => [id]} = FakeHost.await_request()
    assert id == token.id
  end

  # ── Callbacks ─────────────────────────────────────────────────────────────

  test "host invokeCallback runs registered callback and returns result" do
    host = start_host()
    transport = start_transport(host)

    callback_id = Transport.register_callback(transport, fn value -> "got:" <> value end)
    assert is_binary(callback_id)

    invoke_callback(host, 100, callback_id, %{"p0" => "hello"})

    assert_receive {:fake_host, :request, %{"id" => 100, "result" => "got:hello"}}, 1000
  end

  test "callback error is returned as JSON-RPC error" do
    host = start_host()
    transport = start_transport(host)

    callback_id = Transport.register_callback(transport, fn _ -> raise "boom" end)
    invoke_callback(host, 101, callback_id, %{"p0" => "hello"})

    assert_receive {:fake_host, :request, %{"id" => 101, "error" => error}}, 1000
    assert error["message"] =~ "boom"
    refute Map.has_key?(error, "result")
  end

  test "callback that invokes a capability does not deadlock" do
    handler = fn
      %{"method" => "invokeCapability"} -> {:result, %{"$handle" => "7", "$type" => "T"}}
      _ -> :noreply
    end

    host = start_host(handler: handler)
    transport = start_transport(host)

    callback_id =
      Transport.register_callback(transport, fn _context ->
        {:ok, %Handle{id: id}} =
          Transport.invoke_capability(transport, "Aspire.Hosting/build", %{})

        id
      end)

    invoke_callback(host, 102, callback_id, %{"p0" => %{"$handle" => "5", "$type" => "Context"}})

    assert_receive {:fake_host, :request, %{"method" => "invokeCapability"}}, 1000
    assert_receive {:fake_host, :request, %{"id" => 102, "result" => "7"}}, 1000
  end

  test "callback receives a cancellation token for $cancellationToken" do
    host = start_host()
    transport = start_transport(host)

    parent = self()

    callback_id =
      Transport.register_callback(transport, fn context, token ->
        send(parent, {:callback_args, context, token})
        "done"
      end)

    invoke_callback(host, 103, callback_id, %{
      "p0" => %{"$handle" => "5", "$type" => "Context"},
      "p1" => "ct_1",
      "$cancellationToken" => "ct_1"
    })

    assert_receive {:callback_args, %Handle{id: "5"}, %CancellationToken{id: "ct_1"}}, 1000
    assert_receive {:fake_host, :request, %{"id" => 103, "result" => "done"}}, 1000
  end

  # ── Lifecycle ─────────────────────────────────────────────────────────────

  test "socket close stops the transport" do
    host = start_host()
    transport = start_transport(host)

    reference = Process.monitor(transport)
    pending = Task.async(fn -> Transport.invoke_capability(transport, "slow", %{}) end)
    FakeHost.await_request()

    FakeHost.close(host)

    assert {:error, %Aspire.Error{code: "CONNECTION_CLOSED"}} = Task.await(pending)
    assert_receive {:DOWN, ^reference, :process, ^transport, :normal}, 1000
  end

  test "missing REMOTE_APP_HOST_SOCKET_PATH returns a clear error" do
    previous = System.get_env(@socket_env)
    System.delete_env(@socket_env)
    on_exit(fn -> if previous, do: System.put_env(@socket_env, previous) end)

    assert {:error, %Aspire.Error{} = error} = Transport.start_link()
    assert error.code == "MISSING_SOCKET_PATH"
    assert error.message =~ @socket_env
  end

  test "connect reads the socket path from the environment" do
    host = start_host(handler: method_handler(%{"ping" => "pong"}))
    previous = System.get_env(@socket_env)
    System.put_env(@socket_env, FakeHost.socket_path(host))

    on_exit(fn ->
      if previous, do: System.put_env(@socket_env, previous), else: System.delete_env(@socket_env)
    end)

    assert {:ok, transport} = Transport.connect()
    on_exit(fn -> stop_transport(transport) end)

    assert {:ok, "pong"} = Transport.ping()
  end

  test "connect authenticates when ASPIRE_REMOTE_APPHOST_TOKEN is set" do
    host = start_host(handler: method_handler(%{"authenticate" => true}))
    previous = System.get_env(@token_env)
    System.put_env(@token_env, "secret-token")

    on_exit(fn ->
      if previous, do: System.put_env(@token_env, previous), else: System.delete_env(@token_env)
    end)

    transport = start_transport(host)
    assert is_pid(transport)

    assert %{"method" => "authenticate", "params" => ["secret-token"]} = FakeHost.await_request()
  end

  # ── Helpers ───────────────────────────────────────────────────────────────

  defp start_host(opts \\ []) do
    {:ok, host} = FakeHost.start_link(opts)
    path = FakeHost.socket_path(host)
    on_exit(fn -> File.rm(path) end)
    host
  end

  defp start_transport(host, opts \\ []) do
    path = FakeHost.socket_path(host)
    {:ok, transport} = Transport.start_link(Keyword.put(opts, :socket_path, path))
    on_exit(fn -> stop_transport(transport) end)
    transport
  end

  defp stop_transport(transport) do
    if Process.alive?(transport), do: GenServer.stop(transport, :normal)
  catch
    :exit, _ -> :ok
  end

  defp method_handler(results) do
    fn message ->
      case Map.fetch(results, message["method"]) do
        {:ok, value} -> {:result, value}
        :error -> :noreply
      end
    end
  end

  defp invoke_callback(host, id, callback_id, args) do
    FakeHost.send_message(host, %{
      "jsonrpc" => "2.0",
      "id" => id,
      "method" => "invokeCallback",
      "params" => [callback_id, args]
    })
  end

  defp parse_frame("Content-Length: " <> rest) do
    [length, body] = :binary.split(rest, "\r\n\r\n")
    {String.to_integer(length), body}
  end
end

File: tests/Aspire.Hosting.CodeGeneration.Dart.DartTests/test/aspire_transport_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';

import '../../../src/Aspire.Hosting.CodeGeneration.Dart/Resources/base.dart';
import '../../../src/Aspire.Hosting.CodeGeneration.Dart/Resources/transport.dart';
import 'support/fake_host.dart';

void main() {
  // ── Framing ────────────────────────────────────────────────────────────────

  test('frames a request with Content-Length byte count', () async {
    final FakeHost host = await startHost();
    final AspireTransport transport = await startTransport(host);

    final Future<String> pending = transport.ping();
    final Map<String, Object?> request = await host.awaitRequest();
    expect(request['method'], 'ping');

    final _Frame frame = parseFrame(host.receivedBytes);
    expect(frame.length, frame.body.length);

    final Map<String, Object?> decoded =
        jsonDecode(utf8.decode(frame.body)) as Map<String, Object?>;
    expect(decoded['jsonrpc'], '2.0');
    expect(decoded['method'], 'ping');
    expect(decoded['params'], isEmpty);
    expect(decoded['id'], isA<int>());

    await host.sendMessage(<String, Object?>{
      'jsonrpc': '2.0',
      'id': request['id'],
      'result': 'pong',
    });
    expect(await pending, 'pong');
  });

  test('parses two frames in one packet', () async {
    final FakeHost host = await startHost();
    final AspireTransport transport = await startTransport(host);

    final Future<Object?> first = transport.invokeCapability(
      'one',
      <String, Object?>{},
    );
    final Map<String, Object?> firstRequest = await host.awaitRequest();

    final Future<Object?> second = transport.invokeCapability(
      'two',
      <String, Object?>{},
    );
    final Map<String, Object?> secondRequest = await host.awaitRequest();

    await host.sendRaw(<int>[
      ...FakeHost.frame(<String, Object?>{
        'jsonrpc': '2.0',
        'id': firstRequest['id'],
        'result': 'first',
      }),
      ...FakeHost.frame(<String, Object?>{
        'jsonrpc': '2.0',
        'id': secondRequest['id'],
        'result': 'second',
      }),
    ]);

    expect(await first, 'first');
    expect(await second, 'second');
  });

  test('parses a frame split across packets', () async {
    final FakeHost host = await startHost();
    final AspireTransport transport = await startTransport(host);

    final Future<Object?> pending = transport.invokeCapability(
      'one',
      <String, Object?>{},
    );
    final Map<String, Object?> request = await host.awaitRequest();

    bool completed = false;
    unawaited(pending.then((Object? _) => completed = true));

    final List<int> frame = FakeHost.frame(<String, Object?>{
      'jsonrpc': '2.0',
      'id': request['id'],
      'result': 'split',
    });

    await host.sendRaw(frame.sublist(0, 12));
    await settle();
    expect(completed, isFalse);

    await host.sendRaw(frame.sublist(12, 32));
    await settle();
    expect(completed, isFalse);

    await host.sendRaw(frame.sublist(32));
    expect(await pending, 'split');
  });

  test('multibyte UTF-8 body length is bytes not graphemes', () async {
    final FakeHost host = await startHost();
    final AspireTransport transport = await startTransport(host);

    const String text = 'héllo → 世界';
    final Future<Object?> pending = transport.invokeCapability(
      'echo',
      <String, Object?>{'text': text},
    );
    final Map<String, Object?> request = await host.awaitRequest();

    final _Frame frame = parseFrame(host.receivedBytes);
    expect(frame.length, frame.body.length);

    final String body = utf8.decode(frame.body);
    expect(frame.length, greaterThan(body.length));

    final Map<String, Object?> decoded =
        jsonDecode(body) as Map<String, Object?>;
    expect(decoded['params'], <Object?>[
      'echo',
      <String, Object?>{'text': text},
    ]);

    await host.sendMessage(<String, Object?>{
      'jsonrpc': '2.0',
      'id': request['id'],
      'result': text,
    });
    expect(await pending, text);
  });

  // ── Requests ───────────────────────────────────────────────────────────────

  test('ping returns pong', () async {
    final FakeHost host = await startHost(
      handler: methodHandler(<String, Object?>{'ping': 'pong'}),
    );
    final AspireTransport transport = await startTransport(host);

    expect(await transport.ping(), 'pong');

    final Map<String, Object?> request = await host.awaitRequest();
    expect(request['method'], 'ping');
    expect(request['params'], isEmpty);
  });

  test('invokeCapability returns a handle', () async {
    const String redisType =
        'Aspire.Hosting.Redis/Aspire.Hosting.ApplicationModel.RedisResource';
    const String builderType =
        'Aspire.Hosting/Aspire.Hosting.IDistributedApplicationBuilder';

    final FakeHost host = await startHost(
      handler: methodHandler(<String, Object?>{
        'invokeCapability': <String, Object?>{
          r'$handle': '2',
          r'$type': redisType,
        },
      }),
    );
    final AspireTransport transport = await startTransport(host);

    final Object? result = await transport.invokeCapability(
      'Aspire.Hosting.Redis/addRedis',
      <String, Object?>{
        'builder': const AspireHandle('1', builderType),
        'name': 'cache',
      },
    );

    expect(result, isA<AspireHandle>());
    final AspireHandle handle = result! as AspireHandle;
    expect(handle.id, '2');
    expect(handle.type, redisType);

    final Map<String, Object?> request = await host.awaitRequest();
    expect(request['method'], 'invokeCapability');
    expect(request['params'], <Object?>[
      'Aspire.Hosting.Redis/addRedis',
      <String, Object?>{
        'builder': <String, Object?>{r'$handle': '1', r'$type': builderType},
        'name': 'cache',
      },
    ]);
  });

  test(
    'invokeCapability error maps to AspireError with code and data',
    () async {
      final Map<String, Object?> atsError = <String, Object?>{
        'code': 'CAPABILITY_NOT_FOUND',
        'message': 'Unknown capability: Contoso.Aspire/bar',
        'capability': 'Contoso.Aspire/bar',
        'details': <String, Object?>{'parameter': 'name'},
      };

      final FakeHost host = await startHost(
        handler: methodHandler(<String, Object?>{
          'invokeCapability': <String, Object?>{r'$error': atsError},
        }),
      );
      final AspireTransport transport = await startTransport(host);

      await expectLater(
        transport.invokeCapability('Contoso.Aspire/bar', <String, Object?>{}),
        throwsA(
          isA<AspireError>()
              .having(
                (AspireError error) => error.code,
                'code',
                'CAPABILITY_NOT_FOUND',
              )
              .having(
                (AspireError error) => error.message,
                'message',
                'Unknown capability: Contoso.Aspire/bar',
              )
              .having(
                (AspireError error) => error.capability,
                'capability',
                'Contoso.Aspire/bar',
              )
              .having((AspireError error) => error.data, 'data', atsError),
        ),
      );
    },
  );

  test('concurrent requests correlate by id', () async {
    final FakeHost host = await startHost();
    final AspireTransport transport = await startTransport(host);

    final List<Future<Object?>> pending = <Future<Object?>>[
      for (int index = 1; index <= 5; index++)
        transport.invokeCapability('cap$index', <String, Object?>{}),
    ];

    final List<Map<String, Object?>> requests = <Map<String, Object?>>[
      for (int index = 0; index < 5; index++) await host.awaitRequest(),
    ];

    final List<int> ids = requests
        .map((Map<String, Object?> request) => request['id']! as int)
        .toList();
    expect(ids, orderedEquals(<int>[...ids]..sort()));

    // Answer in reverse order to prove correlation is by id, not by arrival.
    for (final Map<String, Object?> request in requests.reversed) {
      final List<Object?> params = request['params']! as List<Object?>;
      final String index = (params[0]! as String).substring('cap'.length);
      await host.sendMessage(<String, Object?>{
        'jsonrpc': '2.0',
        'id': request['id'],
        'result': 'result$index',
      });
    }

    expect(await Future.wait(pending), <Object?>[
      'result1',
      'result2',
      'result3',
      'result4',
      'result5',
    ]);
  });

  test('cancelToken sends cancelToken', () async {
    final FakeHost host = await startHost(
      handler: methodHandler(<String, Object?>{'cancelToken': true}),
    );
    final AspireTransport transport = await startTransport(host);

    expect(await transport.cancelToken('ct_abc'), isTrue);

    final Map<String, Object?> request = await host.awaitRequest();
    expect(request['method'], 'cancelToken');
    expect(request['params'], <Object?>['ct_abc']);
  });

  test('CancellationToken.cancel sends cancelToken', () async {
    final FakeHost host = await startHost(
      handler: methodHandler(<String, Object?>{'cancelToken': true}),
    );
    final AspireTransport transport = await startTransport(host);

    final CancellationToken token = CancellationToken.create();
    expect(await token.cancel(transport), isTrue);

    final Map<String, Object?> request = await host.awaitRequest();
    expect(request['method'], 'cancelToken');
    expect(request['params'], <Object?>[token.id]);
  });

  // ── Callbacks ──────────────────────────────────────────────────────────────

  test(
    'host invokeCallback runs registered callback and returns result',
    () async {
      final FakeHost host = await startHost();
      final AspireTransport transport = await startTransport(host);

      final String callbackId = transport.registerCallback(
        (String value) => 'got:$value',
      );
      expect(callbackId, startsWith('callback_1_'));

      await invokeCallback(host, 100, callbackId, <String, Object?>{
        'p0': 'hello',
      });

      final Map<String, Object?> response = await host.awaitRequest();
      expect(response['id'], 100);
      expect(response['result'], 'got:hello');
    },
  );

  test('callback error is returned as JSON-RPC error', () async {
    final FakeHost host = await startHost();
    final AspireTransport transport = await startTransport(host);

    final String callbackId = transport.registerCallback(
      (String _) => throw StateError('boom'),
    );
    await invokeCallback(host, 101, callbackId, <String, Object?>{
      'p0': 'hello',
    });

    final Map<String, Object?> response = await host.awaitRequest();
    expect(response['id'], 101);
    expect(response.containsKey('result'), isFalse);
    final Map<String, Object?> error =
        response['error']! as Map<String, Object?>;
    expect(error['message'], contains('boom'));
    expect(error['code'], -32000);
  });

  test('callback that invokes a capability does not deadlock', () async {
    final FakeHost host = await startHost(
      handler: methodHandler(<String, Object?>{
        'invokeCapability': <String, Object?>{r'$handle': '7', r'$type': 'T'},
      }),
    );
    final AspireTransport transport = await startTransport(host);

    final String callbackId = transport.registerCallback((Object? _) async {
      final Object? built = await transport.invokeCapability(
        'Aspire.Hosting/build',
        <String, Object?>{},
      );
      return (built! as AspireHandle).id;
    });

    await invokeCallback(host, 102, callbackId, <String, Object?>{
      'p0': <String, Object?>{r'$handle': '5', r'$type': 'Context'},
    });

    final Map<String, Object?> nested = await host.awaitRequest();
    expect(nested['method'], 'invokeCapability');

    final Map<String, Object?> response = await host.awaitRequest();
    expect(response['id'], 102);
    expect(response['result'], '7');
  });

  test(
    r'callback receives a cancellation token for $cancellationToken',
    () async {
      final FakeHost host = await startHost();
      final AspireTransport transport = await startTransport(host);

      final List<Object?> seen = <Object?>[];
      final String callbackId = transport.registerCallback((
        Object? context,
        Object? token,
      ) {
        seen
          ..add(context)
          ..add(token);
        return 'done';
      });

      await invokeCallback(host, 103, callbackId, <String, Object?>{
        'p0': <String, Object?>{r'$handle': '5', r'$type': 'Context'},
        'p1': 'ct_1',
        r'$cancellationToken': 'ct_1',
      });

      final Map<String, Object?> response = await host.awaitRequest();
      expect(response['id'], 103);
      expect(response['result'], 'done');

      expect(seen[0], const AspireHandle('5', 'Context'));
      expect(seen[1], const CancellationToken('ct_1'));
    },
  );

  // ── Lifecycle ──────────────────────────────────────────────────────────────

  test('socket close fails pending requests and completes onClose', () async {
    final FakeHost host = await startHost();
    final AspireTransport transport = await startTransport(host);

    final Future<Object?> pending = transport.invokeCapability(
      'slow',
      <String, Object?>{},
    );
    // Attach the expectation before the close, so the failure never reaches
    // the zone as an unhandled asynchronous error.
    final Future<void> failure = expectLater(
      pending,
      throwsA(
        isA<AspireError>().having(
          (AspireError error) => error.code,
          'code',
          'CONNECTION_CLOSED',
        ),
      ),
    );
    await host.awaitRequest();

    await host.close();

    await failure;
    await transport.onClose;
    expect(transport.isConnected, isFalse);
  });

  test('missing REMOTE_APP_HOST_SOCKET_PATH returns a clear error', () async {
    expect(
      Platform.environment.containsKey(
        AspireTransport.socketPathEnvironmentVariable,
      ),
      isFalse,
      reason:
          'unset ${AspireTransport.socketPathEnvironmentVariable} before you run the test suite',
    );

    await expectLater(
      AspireTransport.connect(),
      throwsA(
        isA<AspireError>()
            .having(
              (AspireError error) => error.code,
              'code',
              'MISSING_SOCKET_PATH',
            )
            .having(
              (AspireError error) => error.message,
              'message',
              contains(AspireTransport.socketPathEnvironmentVariable),
            ),
      ),
    );
  });
}

// ── Helpers ──────────────────────────────────────────────────────────────────

/// Starts a fake host and closes it when the test ends.
Future<FakeHost> startHost({FakeHandler? handler}) async {
  final FakeHost host = await FakeHost.start(handler: handler);
  addTearDown(host.close);
  return host;
}

/// Connects a transport to [host] and closes it when the test ends.
Future<AspireTransport> startTransport(FakeHost host) async {
  final AspireTransport transport = await AspireTransport.connect(
    socketPath: host.socketPath,
  );
  addTearDown(transport.close);
  return transport;
}

/// Sends an `invokeCallback` request from the host to the guest.
Future<void> invokeCallback(
  FakeHost host,
  int id,
  String callbackId,
  Map<String, Object?> args,
) => host.sendMessage(<String, Object?>{
  'jsonrpc': '2.0',
  'id': id,
  'method': 'invokeCallback',
  'params': <Object?>[callbackId, args],
});

/// Waits long enough for a pending socket read to run.
Future<void> settle() => Future<void>.delayed(const Duration(milliseconds: 50));

/// One LSP frame: the declared length and the body bytes.
class _Frame {
  const _Frame(this.length, this.body);

  final int length;
  final List<int> body;
}

/// Reads the first frame out of raw bytes.
_Frame parseFrame(List<int> raw) {
  final String text = latin1.decode(raw);
  final int separator = text.indexOf('\r\n\r\n');
  expect(separator, greaterThan(0), reason: 'the bytes hold no frame header');

  final String headers = text.substring(0, separator);
  expect(headers, startsWith('Content-Length: '));

  final int length = int.parse(
    headers.substring('Content-Length: '.length).trim(),
  );
  final int start = separator + 4;
  return _Frame(length, raw.sublist(start, start + length));
}
